lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -166,6 +168,92 @@ static bool isZeroBitType(Type type) {
   return false;
 }
 
+static Type stripUnpackedTypes(Type type) {
+  return TypeSwitch<Type, Type>(type)
+      .Case<InOutType>([](InOutType inoutType) {
+        return stripUnpackedTypes(inoutType.getElementType());
+      })
+      .Case<UnpackedArrayType>([](UnpackedArrayType arrayType) {
+        return stripUnpackedTypes(arrayType.getElementType());
+      })
+      .Default([](Type type) { return type; });
+}
+
+/// Output the basic type that consists of packed and primitive types.  This is
+/// those to the left of the name in verilog.
+static void printPackedTypeImpl(Type type, raw_ostream &os, Location loc,
+                                SmallVectorImpl<size_t> &dims,
+                                bool implicitIntType, char structFieldSep) {
+  return TypeSwitch<Type, void>(type)
+      .Case<IntegerType>([&](IntegerType integerType) {
+        if (!implicitIntType)
+          os << "logic";
+        if (integerType.getWidth() != 1)
+          dims.push_back(integerType.getWidth());
+        if (!dims.empty() && !implicitIntType)
+          os << ' ';
+
+        for (auto dim : dims)
+          if (dim)
+            os << '[' << (dim - 1) << ":0]";
+          else
+            os << "/*Zero Width*/";
+      })
+      .Case<InOutType>([&](InOutType inoutType) {
+        printPackedTypeImpl(inoutType.getElementType(), os, loc, dims,
+                            implicitIntType, structFieldSep);
+      })
+      .Case<StructType>([&](StructType structType) {
+        os << "struct packed {";
+        for (auto &element : structType.getElements()) {
+          SmallVector<size_t, 8> structDims;
+          printPackedTypeImpl(stripUnpackedTypes(element.type), os, loc,
+                              structDims, false, structFieldSep);
+          os << ' ' << element.name << ';' << structFieldSep;
+        }
+        os << '}';
+      })
+      .Case<ArrayType>([&](ArrayType arrayType) {
+        dims.push_back(arrayType.getSize());
+        printPackedTypeImpl(arrayType.getElementType(), os, loc, dims,
+                            implicitIntType, structFieldSep);
+      })
+      .Case<InterfaceType>([](InterfaceType ifaceType) {
+        // Noop
+      })
+      .Case<UnpackedArrayType>([&](UnpackedArrayType arrayType) {
+        os << "<<unexpected unpacked array>>";
+        emitError(loc, "Unexpected unpacked array in packed type ")
```

Comment:
I think this can be an `assert` since one is supposed to call `stripUnpackedTypes` before calling this, yes?

Reply:
that won't handle packed<unpacked<whatever>>.  Strip will only handle unpacked<packed<whatever>>.

Reply:
Yep. I misread the strip implementation.

Reply:
Yeah this also catches the case if unpacked things end up in structs.